heidrun/hathi_marcxml.rb

```diff
@@ -0,0 +1,350 @@
+  'pd' => 'public domain',
+  'ic-world' => 'in-copyright and permitted as world viewable by the copyright holder',
+  'pdus' => 'public domain only when viewed in the US',
+  'cc-by' => 'Creative Commons Attribution license',
+  'cc-by-nd' => 'Creative Commons Attribution-NoDerivatives license',
+  'cc-by-nc-nd' => 'Creative Commons Attribution-NonCommercial-NoDerivatives license',
+  'cc-by-nc' => 'Creative Commons Attribution-NonCommercial license',
+  'cc-by-nc-sa' => 'Creative Commons Attribution-NonCommercial-ShareAlike license',
+  'cc-by-sa' => 'Creative Commons Attribution-ShareAlike license',
+  'cc-zero' => 'Creative Commons Zero license (implies pd)',
+  'und-world' => 'undetermined copyright status and permitted as world viewable by the depositor',
+}
+
+rights_map = lambda { |r|
+  df_974 = Heidrun::MappingTools::MARC.datafield_els(r, '974')
+  df_974_r = Heidrun::MappingTools::MARC.subfield_values(df_974, 'r')
+
+  df_974_r.map { |rights_name| HATHI_RIGHTS.fetch(rights_name, nil)}
+          .compact
+          .map {|desc| "#{desc.capitalize}. Learn more at http://www.hathitrust.org/access_use"}
+}
+
+subject_map = lambda { |r|
+  datafields = Heidrun::MappingTools::MARC.datafield_els(r, /^6(?:00|[19][0-9]|5[^29])$/)
+
+  subjects = []
+  datafields.each do |df|
+    sfs = Heidrun::MappingTools::MARC.all_subfield_values([df])
+    subjects << sfs.map { |f| f.strip }.join('--')
+  end
+
+  subjects.reject { |s| s.empty? }
+}
+
+title_map = lambda { |r|
+  df_240 = Heidrun::MappingTools::MARC.datafield_els(r, '240')
+  df_240_subs = Heidrun::MappingTools::MARC.all_subfield_values(df_240)
+  df_242 = Heidrun::MappingTools::MARC.datafield_els(r, '242')
+  df_242_subs = Heidrun::MappingTools::MARC.all_subfield_values(df_242)
+  df_245 = Heidrun::MappingTools::MARC.datafield_els(r, '245')
+
+  df_245_subs = df_245.map { |df|
+    df.children.select { |sf|
+      sf.name == 'subfield' && sf[:code] != 'c'
+    }
+  }
+
+  [df_240_subs, df_242_subs, df_245_subs].map{ |sfs| sfs.join(" ")}.reject(&:empty?)
+}
+
+dctype_map = lambda { |r|
+  leader = Heidrun::MappingTools::MARC.leader_value(r)
+  df_337 = Heidrun::MappingTools::MARC.datafield_els(r, '337')
+  df_337a = Heidrun::MappingTools::MARC.subfield_values(df_337, 'a')
+
+  Heidrun::MappingTools::MARC.dctype(:leader => leader,
+                                     :df_337a => df_337a)
+}
+
+fixed_control_fields_map = lambda { |r|
+  Heidrun::MappingTools::MARC.controlfield_values(r, '008')
+}
+
+
+Krikri::Mapper.define(:hathi_marcxml,
+                      :parser => Krikri::MARCXMLParser) do
+
+  # edm:provider
+  #   "HathiTrust Digital Library" (hard coded)
+  provider :class => DPLA::MAP::Agent do
+    uri 'http://mirlyn-aleph.lib.umich.edu/mdp/dpla_metadata/'
+    label 'HathiTrust Digital Library'
+  end
+
+  # edm:dataProvider
+  #   974$u (Map "namespace" to "description,"
+  #   here: https://docs.google.com/document/d/1Bo7ANMMHp4jDmYh6AdlqT3JhM0bkY96lenKnPAFBD_A/edit)
```

Comment:
As per Slack conversation, I'll change this to map from `974$c` against the ht_collections.json entries.